1. What is the relationship between def statements and lambda expressions ?

The def defined functions do not return anything if not explicitly returned whereas the lambda function does return an object. The def functions must be declared in the namespace. The def functions can perform any python task including multiple conditions, nested conditions or loops of any level, printing, importing libraries, raising Exceptions, etc.
on the other hand lambda functions can be used without any declaration in the namespace. The lambda functions defined above are like single-line functions. These functions do not have parenthesis like the def defined functions but instead, take parameters after the lambda keyword as shown above. There is no return keyword defined explicitly because the lambda function does return an object by default.

2. What is the benefit of lambda?

lambda functions can perform limited operations, there is no need of using the return statement and the execution time of the program is fast for the same operation

3. Compare and contrast map, filter, and reduce.

Map operation takes a mapping function and a vector of data as arguments and returns a new vector, which is the result of applying the mapping function on each element of the vector independently. The returned value from map() (map object) then can be passed to functions like list() (to create a list), set() (to create a set) and so on.

The filter function operates on a list and returns a subset of that list after applying the filtering rule.

The reduce function will transform a given list into a single value by applying a given function continuously to all the elements. It basically keeps operating on pairs of elements until there are no more elements left.

4. What are function annotations, and how are they used?

Function annotations provide a way of associating various parts of a function with arbitrary pythoncexpressions at compile time.
Annotations of simple parameters def func(x: expression, y: expression = 20):
Whereas the annotations for excess parameters are as − def func (args: expression, kwargs: expression)

5. What are recursive functions, and how are they used?

A recursive function is a function in code that refers to itself for execution. Recursive functions can be simple or elaborate. They allow for more efficient code writing, for instance, in the listing or compiling of sets of numbers, strings or other variables through a single reiterated process.

In [1]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)
result = factorial(5)
print(result)


120


The function calls factorial(5), which in turn calls factorial(4), then factorial(3), factorial(2), and finally factorial(1). When n becomes 0, the base case is reached, and the recursive calls start unwinding. The result is calculated as 1 * 2 * 3 * 4 * 5, which equals 120.

6. What are some general design guidelines for coding functions?

Use 4-space indentation and no tabs.

Use docstrings

Wrap lines so that they don’t exceed 79 characters

Use of regular and updated comments are valuable to both the coders and users

7. Name three or more ways that functions can communicate results to a caller.

print
return
yield